source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,259 @@
+    }
+
+    ENVOY_LOG(warn, "[Error path] DNS Resolver flushing queries pending callback");
+    parent_.flushPendingQueries();
+
+    // Deallocation of the MainSdRef __has__ to happend __after__ flushing queries. Flushing queries
+    // deallocates individual refs, so deallocating the main ref aheadd would cause deallocation of
+    // invalid individual refs per dns_sd.h
+    parent_.deallocateMainSdRef();
+    // POINT FOR DISCUSSION: can this throw here? Or does it need to be posted like the callback in
+    // flushPendingQueries.
+    parent_.initializeMainSdRef();
+
+    return;
+  }
+
+  // At this point all known non-success cases have been dealt with. Assert that the error code is
+  // no error.
+  // POINT FOR DISCUSSION: Crash otherwise to alert consumers that there are unadressed edge cases?
+  // stat?
+  RELEASE_ASSERT(error_code == kDNSServiceErr_NoError,
+                 "An unknown error has been returned by DNSServiceGetAddrInfoReply");
+
+  // Only add this address to the list if kDNSServiceFlagsAdd is set. Callback targets are purely
+  // additive.
+  if (flags & kDNSServiceFlagsAdd) {
+    auto dns_response = buildDnsResponse(address, ttl);
+    ENVOY_LOG(warn, "Address to add address={}, ttl={}",
+              dns_response.address_->ip()->addressAsString(), ttl);
+
+    if (!pending_cb_) {
+      ENVOY_LOG(warn, "Adding to queries pending callback");
+      pending_cb_ = {ResolutionStatus::Success, {dns_response}};
+      parent_.addPendingQuery(this);
+    } else {
+      ENVOY_LOG(warn, "New address for query already pending flush");
+      pending_cb_->responses_.push_back(dns_response);
+    }
+  }
+
+  if (!(flags & kDNSServiceFlagsMoreComing)) {
+    ENVOY_LOG(warn, "DNS Resolver flushing queries pending callback");
+    parent_.flushPendingQueries();
+  }
+}
+
+DNSServiceErrorType
+AppleDnsResolverImpl::PendingResolution::dnsServiceGetAddrInfo(DnsLookupFamily dns_lookup_family) {
+  DNSServiceProtocol protocol;
+  switch (dns_lookup_family) {
+  case DnsLookupFamily::V4Only:
+    protocol = kDNSServiceProtocol_IPv4;
+    break;
+  case DnsLookupFamily::V6Only:
+    protocol = kDNSServiceProtocol_IPv6;
+    break;
+  case DnsLookupFamily::Auto:
+    protocol = kDNSServiceProtocol_IPv4 | kDNSServiceProtocol_IPv6;
+    break;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+
+  // POINT FOR DISCUSSION: Do we want to allow caching, allow expired answers from the cache?
+  // POINT FOR DISCUSSION: Do we want validation DNSSEC?
+  return DNSServiceGetAddrInfo(
+      &individual_sd_ref_, kDNSServiceFlagsShareConnection | kDNSServiceFlagsTimeout, 0, protocol,
+      dns_name_.c_str(),
+      [](DNSServiceRef, DNSServiceFlags flags, uint32_t interface_index,
+         DNSServiceErrorType error_code, const char* hostname, const struct sockaddr* address,
+         uint32_t ttl, void* context) {
+        static_cast<PendingResolution*>(context)->onDNSServiceGetAddrInfoReply(
```

Comment:
Can you add more comments about the threading in the API? It's not very clear what threads these callbacks can run on.

Reply:
Will do, there is a nice threading comment in the apple docs, will put here.

Reply:
But the short is that the callbacks are ran on the same thread as the dispatcher's that created the FileEvent. So the same as the c-ares impl:

> All calls and callbacks are assumed to happen on the thread that owns the creating dispatcher.